<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Изучение-данных" data-toc-modified-id="Изучение-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Изучение данных</a></span></li><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение и проверка модели</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков" data-toc-modified-id="Расчёт-прибыли-и-рисков-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Расчёт прибыли и рисков</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Выбор локации для скважины

## Загрузка и подготовка данных

### Изучение данных

In [1]:
import pandas as pd, numpy as np #импортируем библиотеки pandas
from sklearn.model_selection import train_test_split #импортируем функцию train_test_split
from sklearn.linear_model import LinearRegression #импортируем библиотеку LinearRegression
from sklearn.metrics import mean_squared_error #импортируем функцию mean_squared_error
import warnings #импортируем библиотеку warnings
warnings.filterwarnings('ignore')

In [2]:
geo_data_0 = pd.read_csv('/datasets/geo_data_0.csv') #считаем файл с данными и сохраним в geo_data_0
geo_data_1 = pd.read_csv('/datasets/geo_data_1.csv') #считаем файл с данными и сохраним в geo_data_1
geo_data_2 = pd.read_csv('/datasets/geo_data_2.csv') #считаем файл с данными и сохраним в geo_data_2

In [3]:
def get_info(dataset):
    display(dataset.head(10)) #получим первые 10 строк таблицы
    dataset.info() #получим общую информацию о данных в таблице
    display(dataset.describe()) #получим также разброс значений, максимум, минимум, среднее и медиану
    display("Доля пропущенных значений:", dataset.isna().mean()) #получим долю пропущенных значений среди всех данных
    print("Количество дубликатов:", dataset.duplicated().sum()) #получим количество дубликатов

In [4]:
get_info(geo_data_0) #применим функцию для geo_data_0

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
5  wX4Hy  0.969570  0.489775 -0.735383   64.741541
6  tL6pL  0.645075  0.530656  1.780266   49.055285
7  BYPU6 -0.400648  0.808337 -5.624670   72.943292
8  j9Oui  0.643105 -0.551583  2.372141  113.356160
9  OLuZU  2.173381  0.563698  9.441852  127.910945

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

'Доля пропущенных значений:'

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

Количество дубликатов: 0


In [5]:
get_info(geo_data_1) #применим функцию для geo_data_1

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
5  HHckp  -3.327590  -2.205276  3.003647   84.038886
6  h5Ujo -11.142655 -10.133399  4.002382  110.992147
7  muH9x   4.234715  -0.001354  2.004588   53.906522
8  YiRkx  13.355129  -0.332068  4.998647  134.766305
9  jG6Gi   1.069227 -11.025667  4.997844  137.945408

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

'Доля пропущенных значений:'

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

Количество дубликатов: 0


In [6]:
get_info(geo_data_2) #применим функцию для geo_data_2

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746
5  LzZXx -0.758092  0.710691  2.585887   90.222465
6  WBHRv -0.574891  0.317727  1.773745   45.641478
7  XO8fn -1.906649 -2.458350 -0.177097   72.480640
8  ybmQ5  1.776292 -0.279356  3.004156  106.616832
9  OilcN -1.214452 -0.439314  5.922514   52.954532

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

'Доля пропущенных значений:'

id         0.0
f0         0.0
f1         0.0
f2         0.0
product    0.0
dtype: float64

Количество дубликатов: 0


###### Вывод

Предоставлены пробы нефти в трёх регионах: geo_data_0 (регион 0), geo_data_1 (регион 1):

    id — уникальный идентификатор скважины;

    f0, f1, f2 — три признака точек (неважно, что они означают, но сами признаки значимы);

    product — объём запасов в скважине (тыс. баррелей).
В каждой из таблиц 5 столбцов. Типы данных в столбцах — object, float64.

В названии колонок нарушения стиля нет. Количество значений в столбцах одинаковое. Дубликатов нет. 

Обнаружены неинформативные столбцы id. Т.к. суть модели в прогнозе количества сырья на основе характеристик скважины.

### Подготовка данных

Так как столбец id для дальнейшего анализа, обучения модели и расчетов не понадобится, удалим. Для этого напишем функцию удаления столбца в каждой таблице.

In [7]:
def drop_column(datasets):
    for i in range(len(datasets)):
        datasets[i] = datasets[i].drop('id', axis=1) # #удалим столбец 'id'
    return datasets

In [8]:
[geo_data_0, geo_data_1, geo_data_2] = drop_column([geo_data_0, geo_data_1, geo_data_2]) #применим функцию drop_column

Проверим данные на зависимость, чтобы в случае высокой связи исключить какой-либо признак.

In [9]:
geo_data_0.corr() #посчитаем корреляцию 

f0        f1        f2   product
f0       1.000000 -0.440723 -0.003153  0.143536
f1      -0.440723  1.000000  0.001724 -0.192356
f2      -0.003153  0.001724  1.000000  0.483663
product  0.143536 -0.192356  0.483663  1.000000

In [10]:
geo_data_1.corr() #посчитаем корреляцию 

f0        f1        f2   product
f0       1.000000  0.182287 -0.001777 -0.030491
f1       0.182287  1.000000 -0.002595 -0.010155
f2      -0.001777 -0.002595  1.000000  0.999397
product -0.030491 -0.010155  0.999397  1.000000

In [11]:
geo_data_2.corr() #посчитаем корреляцию 

f0        f1        f2   product
f0       1.000000  0.000528 -0.000448 -0.001987
f1       0.000528  1.000000  0.000779 -0.001012
f2      -0.000448  0.000779  1.000000  0.445871
product -0.001987 -0.001012  0.445871  1.000000

Корреляция минимальная, исключать признаки не нужно.

Напишем функцию, в которой создадим переменные с целевым и остальными признаками и разделим датафрейм на обучающую и валидационную выборки в соотношении 75:25%.

In [12]:
def data_preparation(dataset):
    features = dataset.drop(['product'], axis=1) #выделим признаки
    target = dataset['product'] #выделим целевой признак product
    
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=12345) #разделим данные на features_train и features_valid в соотношении 75 на 25
    return features_train, features_valid, target_train, target_valid

In [13]:
features_train_0, features_valid_0, target_train_0, target_valid_0 = data_preparation(geo_data_0) #применим функцию data_preparation для geo_data_0

In [14]:
features_train_1, features_valid_1, target_train_1, target_valid_1 = data_preparation(geo_data_1) #применим функцию data_preparation для geo_data_1

In [15]:
features_train_2, features_valid_2, target_train_2, target_valid_2 = data_preparation(geo_data_2) #применим функцию data_preparation для geo_data_2

Сравним размеры получившихся выборок.

In [16]:
print(features_train_0.shape, features_valid_0.shape)

(75000, 3) (25000, 3)


In [17]:
print(features_train_1.shape, features_valid_1.shape)

(75000, 3) (25000, 3)


In [18]:
print(features_train_2.shape, features_valid_2.shape)

(75000, 3) (25000, 3)


### Вывод

Изучив данные, провели предобработку, которая включала в себя:

    исключение ненужных столбцов,
    исследование зависимостей.

Далее подготовили признаки. А именно:

    определили целевой и остальные признаки,
    разделили данные на обучающую и валидационную выборки в соотношении 75:25.

## Обучение и проверка модели

Напишем функцию для обучения модели.

In [19]:
def model_training(features_train, features_valid, target_train, target_valid, region):
    model = LinearRegression()
    model.fit(features_train, target_train) #обучим модель
    predicted_valid = model.predict(features_valid) #получим предсказания модели
    rmse = mean_squared_error(target_valid, predicted_valid, squared=False) #вычислим rmse
    print("Linear Regression для региона:", region)
    print("Средний запас предсказанного сырья:", predicted_valid.mean())
    print("RMSE =", rmse)
    return predicted_valid #возвратим предсказания

Обучим модель на данных региона 0, рассчитаем средний запас предсказанного сырья и rmse.

In [20]:
predicted_valid_0 = model_training(features_train_0, features_valid_0, target_train_0, target_valid_0, 0) #применим функцию model_training

Linear Regression для региона: 0
Средний запас предсказанного сырья: 92.59256778438035
RMSE = 37.5794217150813


Обучим модель на данных региона 1, рассчитаем средний запас предсказанного сырья и rmse.

In [21]:
predicted_valid_1 = model_training(features_train_1, features_valid_1, target_train_1, target_valid_1, 1) #применим функцию model_training

Linear Regression для региона: 1
Средний запас предсказанного сырья: 68.728546895446
RMSE = 0.893099286775617


Обучим модель на данных региона 2, рассчитаем средний запас предсказанного сырья и rmse.

In [22]:
predicted_valid_2 = model_training(features_train_2, features_valid_2, target_train_2, target_valid_2, 2) #применим функцию model_training

Linear Regression для региона: 2
Средний запас предсказанного сырья: 94.96504596800489
RMSE = 40.02970873393434


### Вывод

Исходя из результатов, наибольший средний запас предсказанного сырья в регионе 2, равен 95 тыс. баррелей, rmse при этом равен 40.

## Подготовка к расчёту прибыли

Подготовим данные к расчета прибыли. Так как в данных присутствуют различные ЕИ, приведем к единообразию - млн. рублей. Бюджет - 10 млрд.руб., в млн - 10000. Доход с каждой единицы продукта 450 тыс.руб., в млн - 0.45.

In [23]:
BUDGET = 10000 #бюджет
REVENUE_PER_UNIT_PRODUCT= 0.45 #доход с каждой единицы продукта

Рассчитаем достаточный объём сырья для безубыточной разработки новой скважины. Для этого бюджет поделим на 200 точек и также поделим на доход с каждой единицы продукта и получим результат в тыс.баррелей.

In [24]:
oil_stock_min = BUDGET / 200 / REVENUE_PER_UNIT_PRODUCT
oil_stock_min # минимальный запас нефти в тыс. баррелей

111.11111111111111

### Вывод

Средний запас предсказанного сырья: 92.59256778438035 в регионе 0

Средний запас предсказанного сырья: 68.728546895446 в регионе 1

Средний запас предсказанного сырья: 94.96504596800489 в регионе 2

Таким образом, средние запасы всех регионов не превышают достаточный объём сырья для безубыточной разработки новой скважины.

## Расчёт прибыли и рисков 

Сперва напишем функцию для приведения predictions_valid к типу Series и сброса индексов target_valid.

In [25]:
def normalize(target_valid, predictions_valid):
    predictions_valid = pd.Series(predictions_valid) #приведем predictions_valid к типу Series
    target_valid = target_valid.reset_index(drop=True) #сбросим индексы target_valid
    return target_valid, predictions_valid

In [26]:
target_valid_0, predicted_valid_0 = normalize(target_valid_0, predicted_valid_0) #применим функцию normalize к target_valid_0, predicted_valid_0

In [27]:
target_valid_1, predicted_valid_1 = normalize(target_valid_1, predicted_valid_1) #применим функцию normalize к target_valid_1, predicted_valid_1

In [28]:
target_valid_2, predicted_valid_2 = normalize(target_valid_2, predicted_valid_2) #применим функцию normalize к target_valid_2, predicted_valid_2

Напишем функцию расчета прибыли для полученного объёма сырья в каждом регионе с 200 скважин.

In [29]:
def revenue(target_valid, predictions_valid, count):
    pred_sorted = predictions_valid.sort_values(ascending=False)# каждую выборку отсортируем в порядке убывания сырья
    selected = target_valid[pred_sorted.index][:count]# определяем значения для отобранных 200 скважин
    revenue = (selected.sum() * REVENUE_PER_UNIT_PRODUCT) - BUDGET # доход в млн. рублях с 200 скважин
    return revenue

In [30]:
revenue_0 = revenue(target_valid_0, predicted_valid_0, 200) #применим функцию revenue к target_valid_0, predicted_valid_0
revenue_0 #прибыль в млн. руб.

3320.8260431398503

In [31]:
revenue_1 = revenue(target_valid_1, predicted_valid_1, 200) #применим функцию revenue к target_valid_1, predicted_valid_1
revenue_1 #прибыль в млн. руб.

2415.086696681512

In [32]:
revenue_2 = revenue(target_valid_2, predicted_valid_2, 200) #применим функцию revenue к target_valid_2, predicted_valid_2
revenue_2 #прибыль в млн. руб.

2710.3499635998323

Применим метод бутстреп для расчета средней прогнозной прибыли, доверительного интервала и риска убытков. Для построения 95%-го доверительного интервала нужно выкинуть 5% экстремальных значений, то есть 2.5% самых больших и 2.5% самых маленьких. Точка, на которой заканчивается 2.5% самых маленьких значений, — 0.025-квантиль.. Точка, с которой начинается 2.5% самых больших значений, — 0.975-квантиль.

In [33]:
state = np.random.RandomState(12345) #указываем random_state для воспроизводимости

def get_result(target, predictions):
    values = []
    for i in range(1000):
        target_subsample = target.sample(n=500, replace=True, random_state=state) #создадим подвыборку из ответов
        probs_subsample = predictions[target_subsample.index] #создадим подвыборку из вероятностей, извлекая индексы target_subsample
        result = revenue(target_subsample, probs_subsample, 200) #применим функцию revenue для target_subsample, probs_subsample, 200
        values.append(result) #добавим приюыль в список

    values = pd.Series(values) #приведем к типу Series
    #определим доверительный интервал
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    mean = values.mean() #найдем среднюю прибыль
    risk_of_loss = (values < 0).mean() * 100 #рассчитаем риск убытков
    
    print(f"Средняя прогнозная прибыль: {mean} млн. руб.")
    print(f"С вероятностью 95% прибыль составит от {lower} до {upper} млн. руб.")
    print(f"Риск убытков: {risk_of_loss}%")

In [34]:
get_result(target_valid_0, predicted_valid_0) #применим функцию get_result к target_valid_0, predicted_valid_0

Средняя прогнозная прибыль: 425.9385269105927 млн. руб.
С вероятностью 95% прибыль составит от -102.09009483793592 до 947.976353358369 млн. руб.
Риск убытков: 6.0%


In [35]:
get_result(target_valid_1, predicted_valid_1) #применим функцию get_result к target_valid_1, predicted_valid_1

Средняя прогнозная прибыль: 518.2594936973252 млн. руб.
С вероятностью 95% прибыль составит от 128.12323143308453 до 953.6129820669079 млн. руб.
Риск убытков: 0.3%


In [36]:
get_result(target_valid_2, predicted_valid_2) #применим функцию get_result к target_valid_2, predicted_valid_2

Средняя прогнозная прибыль: 420.19400534405037 млн. руб.
С вероятностью 95% прибыль составит от -115.85260916001138 до 989.6299398445756 млн. руб.
Риск убытков: 6.2%


### Вывод

В регионе 0:

    Прибыль для полученного объёма сырья составляет 3320.8 млн. руб.
    Средняя прогнозная прибыль: 425.9385269105927 млн. руб.
    С вероятностью 95% прибыль составит от -102.09009483793592 до 947.976353358369 млн. руб.
    Риск убытков: 6.0%

В регионе 1:

    Прибыль для полученного объёма сырья составляет 2415.1 млн. руб.
    Средняя прогнозная прибыль: 518.2594936973252 млн. руб.
    С вероятностью 95% прибыль составит от 128.12323143308453 до 953.6129820669079 млн. руб.
    Риск убытков: 0.3%

В регионе 2:

    Прибыль для полученного объёма сырья составляет 2710.3 млн. руб.
    Средняя прогнозная прибыль: 420.19400534405037 млн. руб.
    С вероятностью 95% прибыль составит от -115.85260916001138 до 989.6299398445756 млн. руб.
    Риск убытков: 6.2%

## Общий вывод

Исходя из результатов, регион, в котором риск убытков не превышает 2.5% является регион 1 с вероятностью убытков 0.3%. Средняя прибыль составляет 518.3 млн. руб. Прибыль для полученного объёма сырья составляет 2415.1 млн. руб.
Таким образом, определили регион, где добыча принесёт наибольшую прибыль это регион 1.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован